presto-hive/src/main/java/io/prestosql/plugin/hive/util/SerDeUtils.java

```diff
@@ -312,11 +321,22 @@ private static long formatDateAsLong(Object object, DateObjectInspector inspecto
         return inspector.getPrimitiveJavaObject(object).toEpochDay();
     }
 
-    private static long formatTimestampAsLong(Object object, TimestampObjectInspector inspector)
+    private static DecodedTimestamp formatTimestamp(TimestampType type, Object object, TimestampObjectInspector inspector)
     {
+        long seconds;
+        int nanos;
+
         if (object instanceof TimestampWritable) {
-            return ((TimestampWritable) object).getTimestamp().getTime() * MICROSECONDS_PER_MILLISECOND;
+            LocalDateTime datetime = ((TimestampWritable) object).getTimestamp().toLocalDateTime();
+            seconds = datetime.toEpochSecond(UTC);
+            nanos = datetime.getNano();
+        }
+        else {
+            Timestamp timestamp = inspector.getPrimitiveJavaObject(object);
+            seconds = timestamp.toEpochSecond();
+            nanos = timestamp.getNanos();
         }
-        return inspector.getPrimitiveJavaObject(object).toEpochMilli() * MICROSECONDS_PER_MILLISECOND;
+
+        return new DecodedTimestamp(seconds, (int) round(nanos, 9 - type.getPrecision()));
```

Comment:
rounding here may move you to next second (ie may return NANOS_PER_SECOND)
this needs to be handled here before invoking the DecodedTimestamp ctor

Reply:
I've put it here, though that check should probably also be in the `PrestoTimestampEncoder`s, or maybe `DecodedTimestamp` should have a separate constructor for that.